9.5/root/usr/share/container-scripts/postgresql/common.sh

```diff
@@ -218,3 +223,149 @@ function wait_for_postgresql_master() {
+    hardlink)
+      optimized=:
+      ;;
+    *)
+      echo >&2 "Unsupported value: \$POSTGRESQL_UPGRADE=$POSTGRESQL_UPGRADE"
+      false
+      ;;
+  esac
+
+  # Ensure $PGDATA_new doesn't exist yet, so we can immediately remove it if
+  # there's some problem.
+  test ! -e "$PGDATA_new"
+
+  # initialize the database
+  info_msg "Initialize new data directory; we will migrate to that."
+  initdb_cmd=( initdb_wrapper "$new_pgengine"/initdb "$PGDATA_new" )
+  eval "\${initdb_cmd[@]} ${POSTGRESQL_UPGRADE_INITDB_OPTIONS-}" || \
+    { rm -rf "$PGDATA_new" ; false ; }
+
+  upgrade_cmd=(
+      "$new_pgengine"/pg_upgrade
+      "--old-bindir=$old_pgengine"
+      "--new-bindir=$new_pgengine"
+      "--old-datadir=$PGDATA"
+      "--new-datadir=$PGDATA_new"
+  )
+
+  # Dangerous --link option, we loose $DATADIR if something goes wrong.
+  ! $optimized || upgrade_cmd+=(--link)
+
+  # User-specififed options for pg_upgrade.
+  eval "upgrade_cmd+=(${POSTGRESQL_UPGRADE_PGUPGRADE_OPTIONS-})"
+
+  # the upgrade
+  info_msg "Starting the pg_upgrade process."
+  "${upgrade_cmd[@]}" || { rm -rf "$PGDATA_new" && false ; }
+
+  # Move the important configuration and remove old data.  This is highly
+  # careless, but we can't do more for this over-automatized process.
+  info_msg "Swap the old and new PGDATA and cleanup."
+  mv "$PGDATA"/*.conf "$PGDATA_new"
+  rm -rf "$PGDATA"
+  mv "$PGDATA_new" "$PGDATA"
+
+  info_msg "Upgrade DONE."
+)
+
+
+# Run right after container startup, when the data volume is already initialized
+# (not initialized by this container run) and thus there exists a chance that
+# the data was generated by incompatible PostgreSQL major version.
+try_pgupgrade ()
+{
+  local versionfile="$PGDATA"/PG_VERSION version upgrade_available
+
+  # This file always exists.
+  test -f "$versionfile"
+  version=$(cat "$versionfile")
+
+  # If we don't support pg_upgrade, skip.
+  test -z "${POSTGRESQL_PREV_VERSION-}" && return 0
+
+  # No need to call pg_upgrade, skip.
+  if test "$POSTGRESQL_VERSION" = "$version"; then
+      # Make _sure_ we have this safety-belt here, otherwise our users would
+      # just specify '-e POSTGRESQL_UPGRADE=hardlink' permanently, even for
+      # re-deployment cases when upgrade is not needed.  Setting such
+      # unfortunate default could mean that pg_upgrade might (after some user
+      # mistake) migrate (or even destruct, especially with --link) the old data
+      # directory with limited rollback options, if any.
+      echo >&2
+      echo >&2 "== WARNING!! =="
+      echo >&2 "PostgreSQL server version matches the datadir PG_VERSION."
+      echo >&2 "The \$POSTGRESQL_UPGRADE makes no sense and you probably"
+      echo >&2 "made some mistake, keeping the variable set you might"
+      echo >&2 "risk a data loss in future!"
+      echo >&2 "==============="
+      echo >&2
+
+      # Exit here, but allow _really explicit_ foot-shot.
+      ${POSTGRESQL_UPGRADE_FORCE-false}
+      return 0
+  fi
+
```

Comment:
If one chooses to use environment variables to indicate an upgrade, it can not be used on openshift deployment config, because there is no way to give a replication controller specific env variables on a reschedule. this method can only  be used with a deployer pod.